cmd/kubeadm/app/cmd/join.go

```diff
@@ -340,10 +349,23 @@ func newJoinData(cmd *cobra.Command, args []string, opt *joinOptions, out io.Wri
 		opt.externalcfg.Discovery.BootstrapToken.APIServerEndpoint = args[0]
 	}
 
-	// if not joining a control plane, unset the ControlPlane object
+	// If not passing --control-plane, unset the ControlPlane object
 	if !opt.controlPlane {
-		if opt.externalcfg.ControlPlane != nil {
-			klog.Warningf("[preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when %s flag is not set.", options.ControlPlane)
+		// Use a defaulted JoinControlPlane object to detect if the user has passed
+		// other control-plane related flags.
+		// This defaulting must match the flag defaulting in addJoinConfigFlags()
+		defaultJCP := kubeadmapiv1beta2.JoinControlPlane{
+			LocalAPIEndpoint: kubeadmapiv1beta2.APIEndpoint{BindPort: kubeadmapiv1beta2.DefaultAPIBindPort},
+		}
```

Comment:
@rosti the above is not maintainable as a pattern.
i'd like to use `kubeadmapiv1beta2.SetDefaults_JoinControlPlane`, where applicable.